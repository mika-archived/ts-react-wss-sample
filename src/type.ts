export type MessageRequest = {
  content: string;
};

export type MessageResponse = {
  content: string;
  timestamp: number;
  username: string;
};

export type JoinRequest = {
  roomId: string;
  username: string;
};

export type LeaveRequest = {};
