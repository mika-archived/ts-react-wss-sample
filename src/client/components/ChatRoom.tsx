import React from "react";
import PropTypes from "prop-types";

import Button from "./controls/Button";
import Layout from "./controls/Layout";
import ChatInputBox from "./ChatInputBox";
import ChatMessage from "./ChatMessage";
import { MessageResponse } from "../../type";

type Props = {
  roomId: string;
  messages: MessageResponse[];
  onSubmit: (message: string) => void;
  onLeave: () => void;
};

const ChatRoom: React.FC<Props> = ({ roomId, messages, onSubmit, onLeave }) => {
  return (
    <Layout direction="vertical">
      <div>
        {/* eslint-disable-next-line react/jsx-one-expression-per-line */}
        <h2>Room ID: {roomId}</h2>
        <p>Messages</p>
      </div>
      <div style={{ flex: 1 }}>
        {messages.map(w => (
          <ChatMessage key={w.timestamp} message={w} />
        ))}
      </div>
      <Layout direction="horizontal">
        <div style={{ flex: 1 }}>
          <ChatInputBox onSubmit={onSubmit} />
        </div>
        <Button role="button" onClick={onLeave}>
          Leave Room
        </Button>
      </Layout>
    </Layout>
  );
};

ChatRoom.propTypes = {
  roomId: PropTypes.string.isRequired,
  messages: PropTypes.arrayOf(
    PropTypes.shape({
      content: PropTypes.string.isRequired,
      timestamp: PropTypes.number.isRequired,
      username: PropTypes.string.isRequired
    }).isRequired
  ).isRequired,
  onSubmit: PropTypes.func.isRequired,
  onLeave: PropTypes.func.isRequired
};

export default ChatRoom;
