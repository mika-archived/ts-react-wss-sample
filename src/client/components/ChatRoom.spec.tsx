import "@testing-library/jest-dom/extend-expect";
import "jest-styled-components";

import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";

import ChatRoom from "./ChatRoom";
import { MessageResponse } from "../../type";

const messages = [
  {
    username: "HinataYui",
    timestamp: 1234567890,
    content: "Angel Beats!"
  } as MessageResponse
];

afterEach(cleanup);

test("does not have changes between snapshot", () => {
  const { asFragment } = render(
    <ChatRoom roomId="example" messages={messages} onSubmit={() => {}} onLeave={() => {}} />
  );

  expect(asFragment()).toMatchSnapshot();
});

test("showing RoomID on header", () => {
  const { getByText } = render(<ChatRoom roomId="example" messages={[]} onSubmit={() => {}} onLeave={() => {}} />);

  expect(getByText("Room ID: example")).not.toBeUndefined();
});

test("fire `onSubmit` event when user filled message and click button", () => {
  const onSubmit = jest.fn();
  const { getByRole, getByText } = render(
    <ChatRoom roomId="example" messages={[]} onSubmit={onSubmit} onLeave={() => {}} />
  );

  fireEvent.change(getByRole("textbox"), { target: { value: "Hello" } });
  fireEvent.click(getByText("Sent"));

  expect(onSubmit).toHaveBeenCalledTimes(1);
});

test("fire `onLeave` event when user clicked `leave` button", () => {
  const onLeave = jest.fn();
  const { getByText } = render(<ChatRoom roomId="example" messages={[]} onSubmit={() => {}} onLeave={onLeave} />);

  fireEvent.click(getByText("Leave Room"));

  expect(onLeave).toHaveBeenCalledTimes(1);
});

test("when provided message(s), show it", () => {
  const { getByText } = render(
    <ChatRoom roomId="example" messages={messages} onSubmit={() => {}} onLeave={() => {}} />
  );
  const component = getByText(/HinataYui.*: Angel Beats!/);
  expect(component).not.toBeUndefined();
});
