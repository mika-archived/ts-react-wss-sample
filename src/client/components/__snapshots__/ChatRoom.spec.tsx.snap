// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not have changes between snapshots 1`] = `
<DocumentFragment>
  .c3 {
  height: 1.9rem;
  padding: 0.25rem 0.6rem;
  color: #ffffff;
  background: #0074d9;
  border: 1px solid #0074d9;
}

.c3:disabled {
  color: #dddddd;
  background: #aaaaaa;
  border-color: #aaaaaa;
}

.c4 {
  height: 1.9rem;
  padding: 0.25rem 0.6rem;
  color: #111111;
  background: #ffffff;
  border: 1px solid #dddddd;
}

.c4:disabled {
  color: #dddddd;
  background: #aaaaaa;
  border-color: #aaaaaa;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c2 {
  height: 1.6rem;
  padding: 2px 8px;
  border: 1px solid #dddddd;
  outline: 0;
}

.c2:focus {
  border: 1px solid #0074d9;
}

<div
    class="c0"
    direction="vertical"
  >
    <div>
      <h2>
        Room ID: example
      </h2>
      <p>
        Messages
      </p>
    </div>
    <div
      style="flex: 1;"
    >
      <p>
        HinataYui (2009/02/14 08:31:30) : Angel Beats!
      </p>
    </div>
    <div
      class="c1"
      direction="horizontal"
    >
      <div
        style="flex: 1;"
      >
        <div
          class="c1"
          direction="horizontal"
        >
          <input
            class="c2"
            role="textbox"
            style="flex: 1;"
            value=""
          />
          <button
            class="c3"
            disabled=""
            role="button"
            style="width: 100px;"
            type="submit"
          >
            Sent
          </button>
        </div>
      </div>
      <button
        class="c4"
        role="button"
      >
        Leave Room
      </button>
    </div>
  </div>
</DocumentFragment>
`;
